// @metaplex-foundation/js/dist/cjs/plugins/candyMachineModule/operations/mintCandyMachine.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var mplCandyMachine = require('@metaplex-foundation/mpl-candy-machine');
var web3_js = require('@solana/web3.js');
var accounts = require('../accounts.cjs');
var asserts = require('../asserts.cjs');
var errors = require('../errors.cjs');
var pdas$1 = require('../pdas.cjs');
var program$1 = require('../program.cjs');
var Operation = require('../../../types/Operation.cjs');
var program = require('../../nftModule/program.cjs');
var pdas = require('../../nftModule/pdas.cjs');
var Amount = require('../../../types/Amount.cjs');
var pdas$2 = require('../../tokenModule/pdas.cjs');
var TransactionBuilder = require('../../../utils/TransactionBuilder.cjs');
var Account = require('../../../types/Account.cjs');

// Operation
// -----------------

const Key = 'MintCandyMachineOperation';
/**
 * @group Operations
 * @category Constructors
 */

const mintCandyMachineOperation = Operation.useOperation(Key);
/**
 * @group Operations
 * @category Types
 */

/**
 * @group Operations
 * @category Handlers
 */
const mintCandyMachineOperationHandler = {
  async handle(operation, metaplex, scope) {
    var _operation$input$paye;

    asserts.assertCanMintCandyMachine(operation.input.candyMachine, (_operation$input$paye = operation.input.payer) !== null && _operation$input$paye !== void 0 ? _operation$input$paye : metaplex.identity());
    const builder = await mintCandyMachineBuilder(metaplex, operation.input);
    scope.throwIfCanceled();
    const output = await builder.sendAndConfirm(metaplex, operation.input.confirmOptions);
    scope.throwIfCanceled();
    let nft;

    try {
      nft = await metaplex.nfts().findByMint({
        mintAddress: output.mintSigner.publicKey,
        tokenAddress: output.tokenAddress
      }).run(scope);
    } catch (error) {
      throw new errors.CandyMachineBotTaxError(metaplex.rpc().getSolanaExporerUrl(output.response.signature), error);
    }

    return {
      nft,
      ...output
    };
  }

}; // -----------------
// Builder
// -----------------

/**
 * @group Transaction Builders
 * @category Inputs
 */

/**
 * @group Transaction Builders
 * @category Constructors
 */
const mintCandyMachineBuilder = async (metaplex, params) => {
  var _params$mintNftInstru;

  const {
    candyMachine,
    payer = metaplex.identity(),
    newMint = web3_js.Keypair.generate(),
    newOwner = metaplex.identity().publicKey,
    newToken,
    tokenProgram,
    associatedTokenProgram,
    tokenMetadataProgram = program.TokenMetadataProgram.publicKey,
    candyMachineProgram = program$1.CandyMachineProgram.publicKey
  } = params;
  const newMetadata = pdas.findMetadataPda(newMint.publicKey, tokenMetadataProgram);
  const newEdition = pdas.findMasterEditionV2Pda(newMint.publicKey, tokenMetadataProgram);
  const candyMachineCreator = pdas$1.findCandyMachineCreatorPda(candyMachine.address, candyMachineProgram);
  const candyMachineCollectionAddress = pdas$1.findCandyMachineCollectionPda(candyMachine.address, candyMachineProgram);
  const candyMachineCollectionAccount = accounts.parseCandyMachineCollectionAccount(await metaplex.rpc().getAccount(candyMachineCollectionAddress));
  const tokenWithMintBuilder = await metaplex.tokens().builders().createTokenWithMint({
    decimals: 0,
    initialSupply: Amount.token(1),
    mint: newMint,
    mintAuthority: payer,
    freezeAuthority: payer.publicKey,
    owner: newOwner,
    token: newToken,
    payer,
    tokenProgram,
    associatedTokenProgram,
    createMintAccountInstructionKey: params.createMintAccountInstructionKey,
    initializeMintInstructionKey: params.initializeMintInstructionKey,
    createAssociatedTokenAccountInstructionKey: params.createAssociatedTokenAccountInstructionKey,
    createTokenAccountInstructionKey: params.createTokenAccountInstructionKey,
    initializeTokenInstructionKey: params.initializeTokenInstructionKey,
    mintTokensInstructionKey: params.mintTokensInstructionKey
  });
  const {
    tokenAddress
  } = tokenWithMintBuilder.getContext();
  const mintNftInstruction = mplCandyMachine.createMintNftInstruction({
    candyMachine: candyMachine.address,
    candyMachineCreator: candyMachineCreator,
    payer: payer.publicKey,
    wallet: candyMachine.walletAddress,
    metadata: newMetadata,
    mint: newMint.publicKey,
    mintAuthority: payer.publicKey,
    updateAuthority: payer.publicKey,
    masterEdition: newEdition,
    tokenMetadataProgram,
    clock: web3_js.SYSVAR_CLOCK_PUBKEY,
    recentBlockhashes: web3_js.SYSVAR_SLOT_HASHES_PUBKEY,
    instructionSysvarAccount: web3_js.SYSVAR_INSTRUCTIONS_PUBKEY
  }, {
    creatorBump: candyMachineCreator.bump
  });

  if (candyMachine.whitelistMintSettings) {
    var _params$whitelistToke;

    const whitelistToken = (_params$whitelistToke = params.whitelistToken) !== null && _params$whitelistToke !== void 0 ? _params$whitelistToke : pdas$2.findAssociatedTokenAccountPda(candyMachine.whitelistMintSettings.mint, payer.publicKey, associatedTokenProgram);
    mintNftInstruction.keys.push({
      pubkey: whitelistToken,
      isWritable: true,
      isSigner: false
    }, {
      pubkey: candyMachine.whitelistMintSettings.mint,
      isWritable: true,
      isSigner: false
    }, {
      pubkey: payer.publicKey,
      isWritable: false,
      isSigner: true
    });
  }

  if (candyMachine.tokenMintAddress) {
    var _params$payerToken;

    const payerToken = (_params$payerToken = params.payerToken) !== null && _params$payerToken !== void 0 ? _params$payerToken : pdas$2.findAssociatedTokenAccountPda(candyMachine.tokenMintAddress, payer.publicKey, associatedTokenProgram);
    mintNftInstruction.keys.push({
      pubkey: payerToken,
      isWritable: true,
      isSigner: false
    }, {
      pubkey: payer.publicKey,
      isWritable: false,
      isSigner: true
    });
  }

  return TransactionBuilder.TransactionBuilder.make().setFeePayer(payer).setContext({
    mintSigner: newMint,
    tokenAddress
  }) // Create token and mint accounts.
  .add(tokenWithMintBuilder) // Create the new NFT.
  .add({
    instruction: mintNftInstruction,
    signers: [payer, newMint],
    key: (_params$mintNftInstru = params.mintNftInstructionKey) !== null && _params$mintNftInstru !== void 0 ? _params$mintNftInstru : 'mintNft'
  }) // Set the collection on the NFT.
  .when(candyMachineCollectionAccount.exists, builder => {
    var _params$setCollection;

    Account.assertAccountExists(candyMachineCollectionAccount);
    const collectionMint = candyMachineCollectionAccount.data.mint;
    const collectionMetadata = pdas.findMetadataPda(collectionMint, tokenMetadataProgram);
    const collectionMasterEdition = pdas.findMasterEditionV2Pda(collectionMint, tokenMetadataProgram);
    const collectionAuthorityRecord = pdas.findCollectionAuthorityRecordPda(collectionMint, candyMachineCollectionAccount.publicKey, tokenMetadataProgram);
    return builder.add({
      instruction: mplCandyMachine.createSetCollectionDuringMintInstruction({
        candyMachine: candyMachine.address,
        metadata: newMetadata,
        payer: payer.publicKey,
        collectionPda: candyMachineCollectionAccount.publicKey,
        tokenMetadataProgram,
        instructions: web3_js.SYSVAR_INSTRUCTIONS_PUBKEY,
        collectionMint: candyMachineCollectionAccount.data.mint,
        collectionMetadata,
        collectionMasterEdition,
        authority: candyMachine.authorityAddress,
        collectionAuthorityRecord
      }),
      signers: [payer],
      key: (_params$setCollection = params.setCollectionInstructionKey) !== null && _params$setCollection !== void 0 ? _params$setCollection : 'setCollection'
    });
  });
};

exports.mintCandyMachineBuilder = mintCandyMachineBuilder;
exports.mintCandyMachineOperation = mintCandyMachineOperation;
exports.mintCandyMachineOperationHandler = mintCandyMachineOperationHandler;
