// @metaplex-foundation/js/dist/esm/plugins/auctionHouseModule/createListing.mjs
import { SYSVAR_INSTRUCTIONS_PUBKEY } from '@solana/web3.js';
import { createAuctioneerSellInstruction, createSellInstruction, createPrintListingReceiptInstruction } from '@metaplex-foundation/mpl-auction-house';
import { findAuctionHouseTradeStatePda, findAuctionHouseProgramAsSignerPda, findAuctioneerPda, findListingReceiptPda } from './pdas.mjs';
import { AUCTIONEER_PRICE } from './constants.mjs';
import { AuctioneerAuthorityRequiredError } from './errors.mjs';
import { useOperation } from '../../types/Operation.mjs';
import { token, lamports, amount } from '../../types/Amount.mjs';
import { findMetadataPda } from '../nftModule/pdas.mjs';
import { findAssociatedTokenAccountPda } from '../tokenModule/pdas.mjs';
import { toPublicKey } from '../../types/PublicKey.mjs';
import { isSigner } from '../../types/Signer.mjs';
import { TransactionBuilder } from '../../utils/TransactionBuilder.mjs';

// Operation
// -----------------

const Key = 'CreateListingOperation';
/**
 * @group Operations
 * @category Constructors
 */

const createListingOperation = useOperation(Key);
/**
 * @group Operations
 * @category Types
 */

/**
 * @group Operations
 * @category Handlers
 */
const createListingOperationHandler = {
  handle: async (operation, metaplex) => {
    return createListingBuilder(metaplex, operation.input).sendAndConfirm(metaplex, operation.input.confirmOptions);
  }
}; // -----------------
// Builder
// -----------------

/**
 * @group Transaction Builders
 * @category Inputs
 */

/**
 * @group Transaction Builders
 * @category Constructors
 */
const createListingBuilder = (metaplex, params) => {
  var _params$tokens, _params$price$basisPo, _params$price, _params$seller, _params$authority, _params$tokenAccount, _params$printReceipt, _params$bookkeeper;

  // Data.
  const auctionHouse = params.auctionHouse;
  const tokens = (_params$tokens = params.tokens) !== null && _params$tokens !== void 0 ? _params$tokens : token(1);
  const priceBasisPoint = params.auctioneerAuthority ? AUCTIONEER_PRICE : (_params$price$basisPo = (_params$price = params.price) === null || _params$price === void 0 ? void 0 : _params$price.basisPoints) !== null && _params$price$basisPo !== void 0 ? _params$price$basisPo : 0;
  const price = auctionHouse.isNative ? lamports(priceBasisPoint) : amount(priceBasisPoint, auctionHouse.treasuryMint.currency);

  if (auctionHouse.hasAuctioneer && !params.auctioneerAuthority) {
    throw new AuctioneerAuthorityRequiredError();
  } // Accounts.


  const seller = (_params$seller = params.seller) !== null && _params$seller !== void 0 ? _params$seller : metaplex.identity();
  const authority = (_params$authority = params.authority) !== null && _params$authority !== void 0 ? _params$authority : auctionHouse.authorityAddress;
  const metadata = findMetadataPda(params.mintAccount);
  const tokenAccount = (_params$tokenAccount = params.tokenAccount) !== null && _params$tokenAccount !== void 0 ? _params$tokenAccount : findAssociatedTokenAccountPda(params.mintAccount, toPublicKey(seller));
  const sellerTradeState = findAuctionHouseTradeStatePda(auctionHouse.address, toPublicKey(seller), auctionHouse.treasuryMint.address, params.mintAccount, price.basisPoints, tokens.basisPoints, tokenAccount);
  const freeSellerTradeState = findAuctionHouseTradeStatePda(auctionHouse.address, toPublicKey(seller), auctionHouse.treasuryMint.address, params.mintAccount, lamports(0).basisPoints, tokens.basisPoints, tokenAccount);
  const programAsSigner = findAuctionHouseProgramAsSignerPda();
  const accounts = {
    wallet: toPublicKey(seller),
    tokenAccount,
    metadata,
    authority: toPublicKey(authority),
    auctionHouse: auctionHouse.address,
    auctionHouseFeeAccount: auctionHouse.feeAccountAddress,
    sellerTradeState,
    freeSellerTradeState,
    programAsSigner
  }; // Args.

  const args = {
    tradeStateBump: sellerTradeState.bump,
    freeTradeStateBump: freeSellerTradeState.bump,
    programAsSignerBump: programAsSigner.bump,
    buyerPrice: price.basisPoints,
    tokenSize: tokens.basisPoints
  }; // Sell Instruction.

  let sellInstruction;

  if (params.auctioneerAuthority) {
    sellInstruction = createAuctioneerSellInstruction({ ...accounts,
      auctioneerAuthority: params.auctioneerAuthority.publicKey,
      ahAuctioneerPda: findAuctioneerPda(auctionHouse.address, params.auctioneerAuthority.publicKey)
    }, args);
  } else {
    sellInstruction = createSellInstruction(accounts, args);
  } // Signers.


  const sellSigners = [seller, authority, params.auctioneerAuthority].filter(input => !!input && isSigner(input)); // Receipt.
  // Since createPrintListingReceiptInstruction can't deserialize createAuctioneerSellInstruction due to a bug
  // Don't print Auctioneer Sell receipt for the time being.

  const shouldPrintReceipt = ((_params$printReceipt = params.printReceipt) !== null && _params$printReceipt !== void 0 ? _params$printReceipt : true) && !params.auctioneerAuthority;
  const bookkeeper = (_params$bookkeeper = params.bookkeeper) !== null && _params$bookkeeper !== void 0 ? _params$bookkeeper : metaplex.identity();
  const receipt = findListingReceiptPda(sellerTradeState);
  return TransactionBuilder.make().setContext({
    sellerTradeState,
    freeSellerTradeState,
    tokenAccount,
    metadata,
    seller: toPublicKey(seller),
    receipt: shouldPrintReceipt ? receipt : null,
    bookkeeper: shouldPrintReceipt ? bookkeeper.publicKey : null,
    price,
    tokens
  }) // Create Listing.
  .add({
    instruction: sellInstruction,
    signers: sellSigners,
    key: 'sell'
  }) // Print the Listing Receipt.
  .when(shouldPrintReceipt, builder => builder.add({
    instruction: createPrintListingReceiptInstruction({
      receipt,
      bookkeeper: bookkeeper.publicKey,
      instruction: SYSVAR_INSTRUCTIONS_PUBKEY
    }, {
      receiptBump: receipt.bump
    }),
    signers: [bookkeeper],
    key: 'printListingReceipt'
  }));
};

export { createListingBuilder, createListingOperation, createListingOperationHandler };
