// @metaplex-foundation/js/dist/esm/plugins/auctionHouseModule/pdas.mjs
import { Buffer } from 'buffer';
import { AuctionHouseProgram } from './program.mjs';
import { Pda } from '../../types/Pda.mjs';

/** @group Pdas */

const findAuctionHousePda = (creator, treasuryMint, programId = AuctionHouseProgram.publicKey) => {
  return Pda.find(programId, [Buffer.from('auction_house', 'utf8'), creator.toBuffer(), treasuryMint.toBuffer()]);
};
/** @group Pdas */

const findAuctioneerPda = (auctionHouse, auctioneerAuthority, programId = AuctionHouseProgram.publicKey) => {
  return Pda.find(programId, [Buffer.from('auctioneer', 'utf8'), auctionHouse.toBuffer(), auctioneerAuthority.toBuffer()]);
};
/** @group Pdas */

const findAuctionHouseProgramAsSignerPda = (programId = AuctionHouseProgram.publicKey) => {
  return Pda.find(programId, [Buffer.from('auction_house', 'utf8'), Buffer.from('signer', 'utf8')]);
};
/** @group Pdas */

const findAuctionHouseFeePda = (auctionHouse, programId = AuctionHouseProgram.publicKey) => {
  return Pda.find(programId, [Buffer.from('auction_house', 'utf8'), auctionHouse.toBuffer(), Buffer.from('fee_payer', 'utf8')]);
};
/** @group Pdas */

const findAuctionHouseTreasuryPda = (auctionHouse, programId = AuctionHouseProgram.publicKey) => {
  return Pda.find(programId, [Buffer.from('auction_house', 'utf8'), auctionHouse.toBuffer(), Buffer.from('treasury', 'utf8')]);
};
/** @group Pdas */

const findAuctionHouseBuyerEscrowPda = (auctionHouse, buyer, programId = AuctionHouseProgram.publicKey) => {
  return Pda.find(programId, [Buffer.from('auction_house', 'utf8'), auctionHouse.toBuffer(), buyer.toBuffer()]);
};
/** @group Pdas */

const findAuctionHouseTradeStatePda = (auctionHouse, wallet, treasuryMint, tokenMint, buyPrice, tokenSize, tokenAccount, programId = AuctionHouseProgram.publicKey) => {
  return Pda.find(programId, [Buffer.from('auction_house', 'utf8'), wallet.toBuffer(), auctionHouse.toBuffer(), ...(tokenAccount ? [tokenAccount.toBuffer()] : []), treasuryMint.toBuffer(), tokenMint.toBuffer(), buyPrice.toArrayLike(Buffer, 'le', 8), tokenSize.toArrayLike(Buffer, 'le', 8)]);
};
/** @group Pdas */

const findListingReceiptPda = (tradeState, programId = AuctionHouseProgram.publicKey) => {
  return Pda.find(programId, [Buffer.from('listing_receipt', 'utf8'), tradeState.toBuffer()]);
};
/** @group Pdas */

const findBidReceiptPda = (tradeState, programId = AuctionHouseProgram.publicKey) => {
  return Pda.find(programId, [Buffer.from('bid_receipt', 'utf8'), tradeState.toBuffer()]);
};
/** @group Pdas */

const findPurchaseReceiptPda = (sellerTradeState, buyerTradeState, programId = AuctionHouseProgram.publicKey) => {
  return Pda.find(programId, [Buffer.from('purchase_receipt', 'utf8'), sellerTradeState.toBuffer(), buyerTradeState.toBuffer()]);
};

export { findAuctionHouseBuyerEscrowPda, findAuctionHouseFeePda, findAuctionHousePda, findAuctionHouseProgramAsSignerPda, findAuctionHouseTradeStatePda, findAuctionHouseTreasuryPda, findAuctioneerPda, findBidReceiptPda, findListingReceiptPda, findPurchaseReceiptPda };
