// @metaplex-foundation/js/dist/esm/plugins/candyMachineModule/operations/createCandyMachine.mjs
import { createInitializeCandyMachineInstruction, createSetCollectionInstruction } from '@metaplex-foundation/mpl-candy-machine';
import { Keypair } from '@solana/web3.js';
import { toCandyMachineInstructionData } from '../models/CandyMachine.mjs';
import { getCandyMachineAccountSizeFromData } from '../helpers.mjs';
import { findCandyMachineCollectionPda } from '../pdas.mjs';
import { CandyMachineProgram } from '../program.mjs';
import { assertSameCurrencies, SOL } from '../../../types/Amount.mjs';
import { findMetadataPda, findMasterEditionV2Pda, findCollectionAuthorityRecordPda } from '../../nftModule/pdas.mjs';
import { TokenMetadataProgram } from '../../nftModule/program.mjs';
import { useOperation } from '../../../types/Operation.mjs';
import { toBigNumber } from '../../../types/BigNumber.mjs';
import { toPublicKey } from '../../../types/PublicKey.mjs';
import { TransactionBuilder } from '../../../utils/TransactionBuilder.mjs';
import { isSigner } from '../../../types/Signer.mjs';
import { ExpectedSignerError } from '../../../errors/SdkError.mjs';

// Operation
// -----------------

const Key = 'CreateCandyMachineOperation';
/**
 * @group Operations
 * @category Constructors
 */

const createCandyMachineOperation = useOperation(Key);
/**
 * @group Operations
 * @category Types
 */

/**
 * @group Operations
 * @category Handlers
 */
const createCandyMachineOperationHandler = {
  async handle(operation, metaplex, scope) {
    const builder = await createCandyMachineBuilder(metaplex, operation.input);
    scope.throwIfCanceled();
    const output = await builder.sendAndConfirm(metaplex, operation.input.confirmOptions);
    scope.throwIfCanceled();
    const candyMachine = await metaplex.candyMachines().findByAddress({
      address: output.candyMachineSigner.publicKey
    }).run(scope);
    return { ...output,
      candyMachine
    };
  }

}; // -----------------
// Builder
// -----------------

/**
 * @group Transaction Builders
 * @category Inputs
 */

/**
 * @group Transaction Builders
 * @category Constructors
 */
const createCandyMachineBuilder = async (metaplex, params) => {
  var _params$candyMachine, _params$payer, _params$authority, _params$collection, _params$wallet, _params$tokenMint, _params$symbol, _params$maxEditionSup, _params$isMutable, _params$retainAuthori, _params$goLiveDate, _params$endSettings, _params$creators, _params$hiddenSetting, _params$whitelistMint, _params$gatekeeper, _params$createAccount, _params$initializeCan;

  const candyMachine = (_params$candyMachine = params.candyMachine) !== null && _params$candyMachine !== void 0 ? _params$candyMachine : Keypair.generate();
  const payer = (_params$payer = params.payer) !== null && _params$payer !== void 0 ? _params$payer : metaplex.identity();
  const authority = (_params$authority = params.authority) !== null && _params$authority !== void 0 ? _params$authority : metaplex.identity();
  const collection = (_params$collection = params.collection) !== null && _params$collection !== void 0 ? _params$collection : null;
  const {
    data,
    wallet,
    tokenMint
  } = toCandyMachineInstructionData(candyMachine.publicKey, { ...params,
    wallet: (_params$wallet = params.wallet) !== null && _params$wallet !== void 0 ? _params$wallet : metaplex.identity().publicKey,
    tokenMint: (_params$tokenMint = params.tokenMint) !== null && _params$tokenMint !== void 0 ? _params$tokenMint : null,
    symbol: (_params$symbol = params.symbol) !== null && _params$symbol !== void 0 ? _params$symbol : '',
    maxEditionSupply: (_params$maxEditionSup = params.maxEditionSupply) !== null && _params$maxEditionSup !== void 0 ? _params$maxEditionSup : toBigNumber(0),
    isMutable: (_params$isMutable = params.isMutable) !== null && _params$isMutable !== void 0 ? _params$isMutable : true,
    retainAuthority: (_params$retainAuthori = params.retainAuthority) !== null && _params$retainAuthori !== void 0 ? _params$retainAuthori : true,
    goLiveDate: (_params$goLiveDate = params.goLiveDate) !== null && _params$goLiveDate !== void 0 ? _params$goLiveDate : null,
    endSettings: (_params$endSettings = params.endSettings) !== null && _params$endSettings !== void 0 ? _params$endSettings : null,
    creators: (_params$creators = params.creators) !== null && _params$creators !== void 0 ? _params$creators : [{
      address: metaplex.identity().publicKey,
      share: 100,
      verified: false
    }],
    hiddenSettings: (_params$hiddenSetting = params.hiddenSettings) !== null && _params$hiddenSetting !== void 0 ? _params$hiddenSetting : null,
    whitelistMintSettings: (_params$whitelistMint = params.whitelistMintSettings) !== null && _params$whitelistMint !== void 0 ? _params$whitelistMint : null,
    gatekeeper: (_params$gatekeeper = params.gatekeeper) !== null && _params$gatekeeper !== void 0 ? _params$gatekeeper : null
  });
  const initializeInstruction = createInitializeCandyMachineInstruction({
    candyMachine: candyMachine.publicKey,
    wallet,
    authority: toPublicKey(authority),
    payer: payer.publicKey
  }, {
    data
  });

  if (tokenMint) {
    initializeInstruction.keys.push({
      pubkey: tokenMint,
      isWritable: false,
      isSigner: false
    });
  } else {
    assertSameCurrencies(params.price, SOL);
  }

  return TransactionBuilder.make().setFeePayer(payer).setContext({
    candyMachineSigner: candyMachine,
    payer,
    wallet,
    authority: toPublicKey(authority),
    creators: data.creators
  }) // Create an empty account for the candy machine.
  .add(await metaplex.system().builders().createAccount({
    payer,
    newAccount: candyMachine,
    space: getCandyMachineAccountSizeFromData(data),
    program: CandyMachineProgram.publicKey,
    instructionKey: (_params$createAccount = params.createAccountInstructionKey) !== null && _params$createAccount !== void 0 ? _params$createAccount : 'createAccount'
  })) // Initialize the candy machine account.
  .add({
    instruction: initializeInstruction,
    signers: [candyMachine, payer],
    key: (_params$initializeCan = params.initializeCandyMachineInstructionKey) !== null && _params$initializeCan !== void 0 ? _params$initializeCan : 'initializeCandyMachine'
  }) // Set the collection.
  .when(!!collection, builder => {
    var _params$setCollection;

    if (!isSigner(authority)) {
      throw new ExpectedSignerError('authority', 'PublicKey', {
        problemSuffix: 'You are trying to create a Candy Machine with a Collection NFT. ' + 'In order for the Collection NFT to be set successfully, you must provide the authority as a Signer.',
        solution: 'Please provide the "authority" parameter as a Signer if you want to set the Collection NFT upon creation. ' + 'Alternatively, you may remove the "collection" parameter to create a Candy Machine without an associated Collection NFT.'
      });
    }

    const collectionMint = collection;
    const metadata = findMetadataPda(collectionMint);
    const edition = findMasterEditionV2Pda(collectionMint);
    const collectionPda = findCandyMachineCollectionPda(candyMachine.publicKey);
    const collectionAuthorityRecord = findCollectionAuthorityRecordPda(collectionMint, collectionPda);
    return builder.add({
      instruction: createSetCollectionInstruction({
        candyMachine: candyMachine.publicKey,
        authority: toPublicKey(authority),
        collectionPda,
        payer: payer.publicKey,
        metadata,
        mint: collectionMint,
        edition,
        collectionAuthorityRecord,
        tokenMetadataProgram: TokenMetadataProgram.publicKey
      }),
      signers: [authority],
      key: (_params$setCollection = params.setCollectionInstructionKey) !== null && _params$setCollection !== void 0 ? _params$setCollection : 'setCollection'
    });
  });
};

export { createCandyMachineBuilder, createCandyMachineOperation, createCandyMachineOperationHandler };
