// @metaplex-foundation/js/dist/esm/plugins/candyMachineModule/asserts.mjs
import { MAX_NAME_LENGTH, MAX_URI_LENGTH } from './constants.mjs';
import { CandyMachineIsFullError, CandyMachineCannotAddAmountError, CandyMachineAddItemConstraintsViolatedError, CandyMachineIsEmptyError, CandyMachineNotLiveError, CandyMachineEndedError } from './errors.mjs';
import { EndSettingType } from '@metaplex-foundation/mpl-candy-machine';
import assert from '../../utils/assert.mjs';
import { toBigNumber } from '../../types/BigNumber.mjs';
import { now } from '../../types/DateTime.mjs';

const assertName = name => {
  assert(name.length <= MAX_NAME_LENGTH, `Candy Machine name too long: ${name} (max ${MAX_NAME_LENGTH})`);
};
const assertUri = uri => {
  assert(uri.length <= MAX_URI_LENGTH, `Candy Machine URI too long: ${uri} (max ${MAX_URI_LENGTH})`);
};
const assertNotFull = (candyMachine, index) => {
  if (candyMachine.itemsAvailable.lte(candyMachine.itemsLoaded)) {
    throw new CandyMachineIsFullError(index, candyMachine.itemsAvailable);
  }
};
const assertNotEmpty = candyMachine => {
  if (candyMachine.itemsRemaining.isZero()) {
    throw new CandyMachineIsEmptyError(candyMachine.itemsAvailable);
  }
};
const assertCanAdd = (candyMachine, index, amount) => {
  if (index.addn(amount).gt(candyMachine.itemsAvailable)) {
    throw new CandyMachineCannotAddAmountError(index, amount, candyMachine.itemsAvailable);
  }
};
const assertAllConfigLineConstraints = configLines => {
  for (let i = 0; i < configLines.length; i++) {
    try {
      assertName(configLines[i].name);
      assertUri(configLines[i].uri);
    } catch (error) {
      throw new CandyMachineAddItemConstraintsViolatedError(toBigNumber(i), configLines[i], {
        cause: error
      });
    }
  }
};
const assertCandyMachineIsLive = candyMachine => {
  var _candyMachine$whiteli, _candyMachine$whiteli2;

  const hasWhitelistPresale = (_candyMachine$whiteli = (_candyMachine$whiteli2 = candyMachine.whitelistMintSettings) === null || _candyMachine$whiteli2 === void 0 ? void 0 : _candyMachine$whiteli2.presale) !== null && _candyMachine$whiteli !== void 0 ? _candyMachine$whiteli : false;

  if (hasWhitelistPresale) {
    return;
  }

  const liveDate = candyMachine.goLiveDate;

  if (!liveDate || liveDate.gte(now())) {
    throw new CandyMachineNotLiveError(liveDate);
  }
};
const assertCandyMachineHasNotEnded = candyMachine => {
  const endSettings = candyMachine.endSettings;

  if (!endSettings) {
    return;
  }

  const hasEndedByAmount = endSettings.endSettingType === EndSettingType.Amount && candyMachine.itemsMinted.gte(endSettings.number);
  const hasEndedByDate = endSettings.endSettingType === EndSettingType.Date && endSettings.date.lt(now());

  if (hasEndedByAmount || hasEndedByDate) {
    throw new CandyMachineEndedError(endSettings);
  }
};
const assertCanMintCandyMachine = (candyMachine, payer) => {
  assertNotEmpty(candyMachine);

  if (candyMachine.authorityAddress.equals(payer.publicKey)) {
    return;
  }

  assertCandyMachineIsLive(candyMachine);
  assertCandyMachineHasNotEnded(candyMachine);
};

export { assertAllConfigLineConstraints, assertCanAdd, assertCanMintCandyMachine, assertCandyMachineHasNotEnded, assertCandyMachineIsLive, assertName, assertNotEmpty, assertNotFull, assertUri };
