// @metaplex-foundation/js/dist/cjs/plugins/candyMachineModule/asserts.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var constants = require('./constants.cjs');
var errors = require('./errors.cjs');
var mplCandyMachine = require('@metaplex-foundation/mpl-candy-machine');
var assert = require('../../utils/assert.cjs');
var BigNumber = require('../../types/BigNumber.cjs');
var DateTime = require('../../types/DateTime.cjs');

const assertName = name => {
  assert["default"](name.length <= constants.MAX_NAME_LENGTH, `Candy Machine name too long: ${name} (max ${constants.MAX_NAME_LENGTH})`);
};
const assertUri = uri => {
  assert["default"](uri.length <= constants.MAX_URI_LENGTH, `Candy Machine URI too long: ${uri} (max ${constants.MAX_URI_LENGTH})`);
};
const assertNotFull = (candyMachine, index) => {
  if (candyMachine.itemsAvailable.lte(candyMachine.itemsLoaded)) {
    throw new errors.CandyMachineIsFullError(index, candyMachine.itemsAvailable);
  }
};
const assertNotEmpty = candyMachine => {
  if (candyMachine.itemsRemaining.isZero()) {
    throw new errors.CandyMachineIsEmptyError(candyMachine.itemsAvailable);
  }
};
const assertCanAdd = (candyMachine, index, amount) => {
  if (index.addn(amount).gt(candyMachine.itemsAvailable)) {
    throw new errors.CandyMachineCannotAddAmountError(index, amount, candyMachine.itemsAvailable);
  }
};
const assertAllConfigLineConstraints = configLines => {
  for (let i = 0; i < configLines.length; i++) {
    try {
      assertName(configLines[i].name);
      assertUri(configLines[i].uri);
    } catch (error) {
      throw new errors.CandyMachineAddItemConstraintsViolatedError(BigNumber.toBigNumber(i), configLines[i], {
        cause: error
      });
    }
  }
};
const assertCandyMachineIsLive = candyMachine => {
  var _candyMachine$whiteli, _candyMachine$whiteli2;

  const hasWhitelistPresale = (_candyMachine$whiteli = (_candyMachine$whiteli2 = candyMachine.whitelistMintSettings) === null || _candyMachine$whiteli2 === void 0 ? void 0 : _candyMachine$whiteli2.presale) !== null && _candyMachine$whiteli !== void 0 ? _candyMachine$whiteli : false;

  if (hasWhitelistPresale) {
    return;
  }

  const liveDate = candyMachine.goLiveDate;

  if (!liveDate || liveDate.gte(DateTime.now())) {
    throw new errors.CandyMachineNotLiveError(liveDate);
  }
};
const assertCandyMachineHasNotEnded = candyMachine => {
  const endSettings = candyMachine.endSettings;

  if (!endSettings) {
    return;
  }

  const hasEndedByAmount = endSettings.endSettingType === mplCandyMachine.EndSettingType.Amount && candyMachine.itemsMinted.gte(endSettings.number);
  const hasEndedByDate = endSettings.endSettingType === mplCandyMachine.EndSettingType.Date && endSettings.date.lt(DateTime.now());

  if (hasEndedByAmount || hasEndedByDate) {
    throw new errors.CandyMachineEndedError(endSettings);
  }
};
const assertCanMintCandyMachine = (candyMachine, payer) => {
  assertNotEmpty(candyMachine);

  if (candyMachine.authorityAddress.equals(payer.publicKey)) {
    return;
  }

  assertCandyMachineIsLive(candyMachine);
  assertCandyMachineHasNotEnded(candyMachine);
};

exports.assertAllConfigLineConstraints = assertAllConfigLineConstraints;
exports.assertCanAdd = assertCanAdd;
exports.assertCanMintCandyMachine = assertCanMintCandyMachine;
exports.assertCandyMachineHasNotEnded = assertCandyMachineHasNotEnded;
exports.assertCandyMachineIsLive = assertCandyMachineIsLive;
exports.assertName = assertName;
exports.assertNotEmpty = assertNotEmpty;
exports.assertNotFull = assertNotFull;
exports.assertUri = assertUri;
