// @metaplex-foundation/js/dist/cjs/types/Account.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var SdkError = require('../errors/SdkError.cjs');

function parseAccount(account, parser) {
  if ('exists' in account && !account.exists) {
    return account;
  }

  return getAccountParsingFunction(parser)(account);
}
function getAccountParsingFunction(parser) {
  function parse(account) {
    if ('exists' in account && !account.exists) {
      return account;
    }

    try {
      const data = parser.deserialize(account.data)[0];
      return { ...account,
        data
      };
    } catch (error) {
      throw new SdkError.UnexpectedAccountError(account.publicKey, parser.name, {
        cause: error
      });
    }
  }

  return parse;
}
function toAccount(account, parser, solution) {
  if ('exists' in account) {
    assertAccountExists(account, parser.name, solution);
  }

  return getAccountParsingFunction(parser)(account);
}
function getAccountParsingAndAssertingFunction(parser) {
  const parse = getAccountParsingFunction(parser);
  return (unparsedAccount, solution) => {
    if ('exists' in unparsedAccount) {
      assertAccountExists(unparsedAccount, parser.name, solution);
    }

    return parse(unparsedAccount);
  };
}
function assertAccountExists(account, name, solution) {
  if (!account.exists) {
    throw new SdkError.AccountNotFoundError(account.publicKey, name, {
      solution
    });
  }
}

exports.assertAccountExists = assertAccountExists;
exports.getAccountParsingAndAssertingFunction = getAccountParsingAndAssertingFunction;
exports.getAccountParsingFunction = getAccountParsingFunction;
exports.parseAccount = parseAccount;
exports.toAccount = toAccount;
