// @metaplex-foundation/js/dist/cjs/plugins/candyMachineModule/models/CandyMachine.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var mplCandyMachine = require('@metaplex-foundation/mpl-candy-machine');
var helpers = require('../helpers.cjs');
var program = require('../program.cjs');
var assert = require('../../../utils/assert.cjs');
var BigNumber = require('../../../types/BigNumber.cjs');
var Amount = require('../../../types/Amount.cjs');
var common = require('../../../utils/common.cjs');
var DateTime = require('../../../types/DateTime.cjs');

// Model
// -----------------

/** @group Models */

// -----------------
// Program to Model
// -----------------

/** @group Model Helpers */
const isCandyMachine = value => typeof value === 'object' && value.model === 'candyMachine';
/** @group Model Helpers */

function assertCandyMachine(value) {
  assert["default"](isCandyMachine(value), 'Expected CandyMachine type');
}
/** @group Model Helpers */

const toCandyMachine = (account, unparsedAccount, collectionAccount) => {
  const itemsAvailable = BigNumber.toBigNumber(account.data.data.itemsAvailable);
  const itemsMinted = BigNumber.toBigNumber(account.data.itemsRedeemed);
  const endSettings = account.data.data.endSettings;
  const hiddenSettings = account.data.data.hiddenSettings;
  const whitelistMintSettings = account.data.data.whitelistMintSettings;
  const gatekeeper = account.data.data.gatekeeper;
  const rawData = unparsedAccount.data;
  const itemsLoaded = hiddenSettings ? BigNumber.toBigNumber(0) : helpers.countCandyMachineItems(rawData);
  const items = hiddenSettings ? [] : helpers.parseCandyMachineItems(rawData);
  return {
    model: 'candyMachine',
    address: account.publicKey,
    programAddress: account.owner,
    version: account.owner.equals(program.CandyMachineProgram.publicKey) ? 2 : 1,
    authorityAddress: account.data.authority,
    walletAddress: account.data.wallet,
    tokenMintAddress: account.data.tokenMint,
    collectionMintAddress: collectionAccount && collectionAccount.exists ? collectionAccount.data.mint : null,
    uuid: account.data.data.uuid,
    price: Amount.lamports(account.data.data.price),
    symbol: common.removeEmptyChars(account.data.data.symbol),
    sellerFeeBasisPoints: account.data.data.sellerFeeBasisPoints,
    isMutable: account.data.data.isMutable,
    retainAuthority: account.data.data.retainAuthority,
    goLiveDate: DateTime.toOptionDateTime(account.data.data.goLiveDate),
    maxEditionSupply: BigNumber.toBigNumber(account.data.data.maxSupply),
    items,
    itemsAvailable,
    itemsMinted,
    itemsRemaining: BigNumber.toBigNumber(itemsAvailable.sub(itemsMinted)),
    itemsLoaded,
    isFullyLoaded: itemsAvailable.lte(itemsLoaded),
    endSettings: endSettings ? endSettings.endSettingType === mplCandyMachine.EndSettingType.Date ? {
      endSettingType: mplCandyMachine.EndSettingType.Date,
      date: DateTime.toDateTime(endSettings.number)
    } : {
      endSettingType: mplCandyMachine.EndSettingType.Amount,
      number: BigNumber.toBigNumber(endSettings.number)
    } : null,
    hiddenSettings,
    whitelistMintSettings: whitelistMintSettings ? { ...whitelistMintSettings,
      discountPrice: whitelistMintSettings.discountPrice ? Amount.lamports(whitelistMintSettings.discountPrice) : null
    } : null,
    gatekeeper: gatekeeper ? { ...gatekeeper,
      network: gatekeeper.gatekeeperNetwork
    } : null,
    creators: account.data.data.creators
  };
}; // -----------------
// Model to Configs
// -----------------

/** @group Models */

/** @group Model Helpers */
const toCandyMachineConfigs = candyMachine => {
  return {
    wallet: candyMachine.walletAddress,
    tokenMint: candyMachine.tokenMintAddress,
    ...candyMachine
  };
}; // -----------------
// Configs to Program
// -----------------

/** @group Models */

/** @group Model Helpers */
const toCandyMachineInstructionData = (address, configs) => {
  var _whitelistMintSetting, _whitelistMintSetting2;

  const endSettings = configs.endSettings;
  const whitelistMintSettings = configs.whitelistMintSettings;
  const gatekeeper = configs.gatekeeper;
  return {
    wallet: configs.wallet,
    tokenMint: configs.tokenMint,
    data: { ...configs,
      uuid: helpers.getCandyMachineUuidFromAddress(address),
      price: configs.price.basisPoints,
      maxSupply: configs.maxEditionSupply,
      endSettings: endSettings ? { ...endSettings,
        number: endSettings.endSettingType === mplCandyMachine.EndSettingType.Date ? endSettings.date : endSettings.number
      } : null,
      whitelistMintSettings: whitelistMintSettings ? { ...whitelistMintSettings,
        discountPrice: (_whitelistMintSetting = (_whitelistMintSetting2 = whitelistMintSettings.discountPrice) === null || _whitelistMintSetting2 === void 0 ? void 0 : _whitelistMintSetting2.basisPoints) !== null && _whitelistMintSetting !== void 0 ? _whitelistMintSetting : null
      } : null,
      gatekeeper: gatekeeper ? { ...gatekeeper,
        gatekeeperNetwork: gatekeeper.network
      } : null
    }
  };
};

exports.assertCandyMachine = assertCandyMachine;
exports.isCandyMachine = isCandyMachine;
exports.toCandyMachine = toCandyMachine;
exports.toCandyMachineConfigs = toCandyMachineConfigs;
exports.toCandyMachineInstructionData = toCandyMachineInstructionData;
