// @metaplex-foundation/js/dist/cjs/plugins/auctionHouseModule/createListing.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var web3_js = require('@solana/web3.js');
var mplAuctionHouse = require('@metaplex-foundation/mpl-auction-house');
var pdas$2 = require('./pdas.cjs');
var constants = require('./constants.cjs');
var errors = require('./errors.cjs');
var Operation = require('../../types/Operation.cjs');
var Amount = require('../../types/Amount.cjs');
var pdas = require('../nftModule/pdas.cjs');
var pdas$1 = require('../tokenModule/pdas.cjs');
var PublicKey = require('../../types/PublicKey.cjs');
var Signer = require('../../types/Signer.cjs');
var TransactionBuilder = require('../../utils/TransactionBuilder.cjs');

// Operation
// -----------------

const Key = 'CreateListingOperation';
/**
 * @group Operations
 * @category Constructors
 */

const createListingOperation = Operation.useOperation(Key);
/**
 * @group Operations
 * @category Types
 */

/**
 * @group Operations
 * @category Handlers
 */
const createListingOperationHandler = {
  handle: async (operation, metaplex) => {
    return createListingBuilder(metaplex, operation.input).sendAndConfirm(metaplex, operation.input.confirmOptions);
  }
}; // -----------------
// Builder
// -----------------

/**
 * @group Transaction Builders
 * @category Inputs
 */

/**
 * @group Transaction Builders
 * @category Constructors
 */
const createListingBuilder = (metaplex, params) => {
  var _params$tokens, _params$price$basisPo, _params$price, _params$seller, _params$authority, _params$tokenAccount, _params$printReceipt, _params$bookkeeper;

  // Data.
  const auctionHouse = params.auctionHouse;
  const tokens = (_params$tokens = params.tokens) !== null && _params$tokens !== void 0 ? _params$tokens : Amount.token(1);
  const priceBasisPoint = params.auctioneerAuthority ? constants.AUCTIONEER_PRICE : (_params$price$basisPo = (_params$price = params.price) === null || _params$price === void 0 ? void 0 : _params$price.basisPoints) !== null && _params$price$basisPo !== void 0 ? _params$price$basisPo : 0;
  const price = auctionHouse.isNative ? Amount.lamports(priceBasisPoint) : Amount.amount(priceBasisPoint, auctionHouse.treasuryMint.currency);

  if (auctionHouse.hasAuctioneer && !params.auctioneerAuthority) {
    throw new errors.AuctioneerAuthorityRequiredError();
  } // Accounts.


  const seller = (_params$seller = params.seller) !== null && _params$seller !== void 0 ? _params$seller : metaplex.identity();
  const authority = (_params$authority = params.authority) !== null && _params$authority !== void 0 ? _params$authority : auctionHouse.authorityAddress;
  const metadata = pdas.findMetadataPda(params.mintAccount);
  const tokenAccount = (_params$tokenAccount = params.tokenAccount) !== null && _params$tokenAccount !== void 0 ? _params$tokenAccount : pdas$1.findAssociatedTokenAccountPda(params.mintAccount, PublicKey.toPublicKey(seller));
  const sellerTradeState = pdas$2.findAuctionHouseTradeStatePda(auctionHouse.address, PublicKey.toPublicKey(seller), auctionHouse.treasuryMint.address, params.mintAccount, price.basisPoints, tokens.basisPoints, tokenAccount);
  const freeSellerTradeState = pdas$2.findAuctionHouseTradeStatePda(auctionHouse.address, PublicKey.toPublicKey(seller), auctionHouse.treasuryMint.address, params.mintAccount, Amount.lamports(0).basisPoints, tokens.basisPoints, tokenAccount);
  const programAsSigner = pdas$2.findAuctionHouseProgramAsSignerPda();
  const accounts = {
    wallet: PublicKey.toPublicKey(seller),
    tokenAccount,
    metadata,
    authority: PublicKey.toPublicKey(authority),
    auctionHouse: auctionHouse.address,
    auctionHouseFeeAccount: auctionHouse.feeAccountAddress,
    sellerTradeState,
    freeSellerTradeState,
    programAsSigner
  }; // Args.

  const args = {
    tradeStateBump: sellerTradeState.bump,
    freeTradeStateBump: freeSellerTradeState.bump,
    programAsSignerBump: programAsSigner.bump,
    buyerPrice: price.basisPoints,
    tokenSize: tokens.basisPoints
  }; // Sell Instruction.

  let sellInstruction;

  if (params.auctioneerAuthority) {
    sellInstruction = mplAuctionHouse.createAuctioneerSellInstruction({ ...accounts,
      auctioneerAuthority: params.auctioneerAuthority.publicKey,
      ahAuctioneerPda: pdas$2.findAuctioneerPda(auctionHouse.address, params.auctioneerAuthority.publicKey)
    }, args);
  } else {
    sellInstruction = mplAuctionHouse.createSellInstruction(accounts, args);
  } // Signers.


  const sellSigners = [seller, authority, params.auctioneerAuthority].filter(input => !!input && Signer.isSigner(input)); // Receipt.
  // Since createPrintListingReceiptInstruction can't deserialize createAuctioneerSellInstruction due to a bug
  // Don't print Auctioneer Sell receipt for the time being.

  const shouldPrintReceipt = ((_params$printReceipt = params.printReceipt) !== null && _params$printReceipt !== void 0 ? _params$printReceipt : true) && !params.auctioneerAuthority;
  const bookkeeper = (_params$bookkeeper = params.bookkeeper) !== null && _params$bookkeeper !== void 0 ? _params$bookkeeper : metaplex.identity();
  const receipt = pdas$2.findListingReceiptPda(sellerTradeState);
  return TransactionBuilder.TransactionBuilder.make().setContext({
    sellerTradeState,
    freeSellerTradeState,
    tokenAccount,
    metadata,
    seller: PublicKey.toPublicKey(seller),
    receipt: shouldPrintReceipt ? receipt : null,
    bookkeeper: shouldPrintReceipt ? bookkeeper.publicKey : null,
    price,
    tokens
  }) // Create Listing.
  .add({
    instruction: sellInstruction,
    signers: sellSigners,
    key: 'sell'
  }) // Print the Listing Receipt.
  .when(shouldPrintReceipt, builder => builder.add({
    instruction: mplAuctionHouse.createPrintListingReceiptInstruction({
      receipt,
      bookkeeper: bookkeeper.publicKey,
      instruction: web3_js.SYSVAR_INSTRUCTIONS_PUBKEY
    }, {
      receiptBump: receipt.bump
    }),
    signers: [bookkeeper],
    key: 'printListingReceipt'
  }));
};

exports.createListingBuilder = createListingBuilder;
exports.createListingOperation = createListingOperation;
exports.createListingOperationHandler = createListingOperationHandler;
